File: sdk/chocobase-dart/lib/chocobase.dart
/// Official Dart and Flutter client for ChocoBase.
library chocobase;

export 'src/auth.dart';
export 'src/postgrest.dart';
export 'src/storage.dart';
export 'src/realtime.dart';
export 'src/functions.dart';

import 'src/auth.dart';
import 'src/postgrest.dart';
import 'src/storage.dart';
import 'src/realtime.dart';
import 'src/functions.dart';

/// Main client for connecting to a ChocoBase project.
class ChocoClient {
  final String url;
  final String apiKey;
  final Map<String, String> headers;

  late final AuthClient auth;
  late final StorageClient storage;
  late final FunctionsClient functions;
  late final RealtimeClient realtime;

  ChocoClient(this.url, this.apiKey, {Map<String, String>? customHeaders})
      : headers = {
          'apikey': apiKey,
          'Authorization': 'Bearer $apiKey',
          'Content-Type': 'application/json',
          ...?customHeaders,
        } {
    auth = AuthClient(url, headers);
    storage = StorageClient(url, headers);
    functions = FunctionsClient(url, headers);
    realtime = RealtimeClient(url, apiKey);
  }

  /// Create a query builder targeting a specific database table.
  QueryBuilder from(String table) {
    return QueryBuilder(url, table, headers);
  }
}

/// Helper function to create a new [ChocoClient] instance.
ChocoClient createClient(String url, String apiKey, {Map<String, String>? customHeaders}) {
  return ChocoClient(url, apiKey, customHeaders: customHeaders);
}

File: sdk/chocobase-swift/Sources/ChocoBase/ChocoBase.swift
import Foundation

/// Official Swift Client for ChocoBase.
public class ChocoClient {
    public let url: URL
    public let apiKey: String
    public let headers: [String: String]
    
    public let auth: AuthClient
    public let storage: StorageClient
    public let functions: FunctionsClient
    
    public init(url: URL, apiKey: String, customHeaders: [String: String]? = nil) {
        self.url = url
        self.apiKey = apiKey
        var h = [
            "apikey": apiKey,
            "Authorization": "Bearer \(apiKey)",
            "Content-Type": "application/json"
        ]
        if let custom = customHeaders {
            for (k, v) in custom {
                h[k] = v
            }
        }
        self.headers = h
        self.auth = AuthClient(url: url, headers: h)
        self.storage = StorageClient(url: url, headers: h)
        self.functions = FunctionsClient(url: url, headers: h)
    }
    
    public func from(_ table: String) -> QueryBuilder {
        return QueryBuilder(baseUrl: url, table: table, headers: headers)
    }
}

public func createClient(url: URL, apiKey: String, customHeaders: [String: String]? = nil) -> ChocoClient {
    return ChocoClient(url: url, apiKey: apiKey, customHeaders: customHeaders)
}
